File: src/main/kotlin/dev/slne/surf/essentialsold/commands/minecraft/DifficultyCommand.java
package dev.slne.surf.essentialsold.commands.minecraft;

import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.jorel.commandapi.wrappers.NativeProxyCommandSender;
import dev.slne.surf.essentialsold.commands.EssentialsCommand;
import dev.slne.surf.essentialsold.utils.EssentialsUtil;
import dev.slne.surf.essentialsold.utils.color.Colors;
import dev.slne.surf.essentialsold.utils.brigadier.Exceptions;
import dev.slne.surf.essentialsold.utils.permission.Permissions;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.Difficulty;

public class DifficultyCommand extends EssentialsCommand {
    public DifficultyCommand() {
        super("difficulty", "difficulty <difficulty>", "Get or change the server difficulty");

        withPermission(Permissions.DIFFICULTY_PERMISSION);

        executesNative((NativeResultingCommandExecutor) (sender, args) -> getDifficulty(sender));
        then(difficultyArgument("difficulty")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> setDifficulty(sender, args.getUnchecked("difficulty"))));
    }


    private static int setDifficulty(NativeProxyCommandSender source, Difficulty difficulty) throws WrapperCommandSyntaxException {
        val world = source.getWorld();

        if (world.getDifficulty() == difficulty) throw Exceptions.ERROR_ALREADY_THAT_DIFFICULT.create(difficulty);
        world.setDifficulty(difficulty);

        EssentialsUtil.sendSuccess(source.getCallee(), Component.text("Die Schwierigkeit wurde auf ", Colors.SUCCESS)
                .append(Component.translatable(difficulty.translationKey(), Colors.VARIABLE_VALUE))
                .append(Component.text(" gesetzt!", Colors.SUCCESS)));

        return 1;
    }

    private static int getDifficulty(NativeProxyCommandSender source) {
        final Difficulty difficulty = source.getWorld().getDifficulty();

        EssentialsUtil.sendSuccess(source.getCallee(), Component.text("Die Schwierigkeit ist ", Colors.INFO)
                .append(Component.translatable(difficulty.translationKey(), Colors.VARIABLE_VALUE))
                .append(Component.text("!", Colors.INFO)));
        return 1;
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/minecraft/DifficultyCommand.kt
package dev.slne.surf.essentials.command.minecraft

import dev.jorel.commandapi.kotlindsl.*
import dev.slne.surf.essentials.command.argument.difficultyArgument
import dev.slne.surf.essentials.util.permission.EssentialsPermissionRegistry
import dev.slne.surf.essentials.util.util.translatable
import dev.slne.surf.surfapi.core.api.messages.Colors
import dev.slne.surf.surfapi.core.api.messages.adventure.sendText
import org.bukkit.Bukkit
import org.bukkit.Difficulty
import org.bukkit.World

fun difficultyCommand() = commandTree("difficulty") {
    withPermission(EssentialsPermissionRegistry.DIFFICULTY_COMMAND)
    playerExecutor { player, _ ->
        player.sendText {
            appendPrefix()
            info("Der Schwierigkeitsgrad der Welt ")
            variableValue(player.world.name)
            info(" ist auf ")
            translatable(player.world.difficulty.translationKey()).colorIfAbsent(Colors.VARIABLE_VALUE)
            info(" gesetzt.")
        }
    }
    difficultyArgument("difficulty") {
        playerExecutor { player, args ->
            val difficulty: Difficulty by args

            Bukkit.getWorlds().forEach { it.difficulty = difficulty }
            player.sendText {
                appendPrefix()
                success("Du hast den Schwierigkeitsgrad aller Welten auf ")
                translatable(difficulty.translationKey()).colorIfAbsent(Colors.VARIABLE_VALUE)
                success(" gesetzt.")
            }
        }

        worldArgument("world") {
            anyExecutor { executor, args ->
                val difficulty: Difficulty by args
                val world: World by args

                if (difficulty == world.difficulty) {
                    executor.sendText {
                        appendPrefix()
                        error("Der Schwierigkeitsgrad der Welt ")
                        variableValue(world.name)
                        error(" ist bereits auf ")
                        translatable(difficulty.translationKey()).colorIfAbsent(Colors.VARIABLE_VALUE)
                        error(" gesetzt.")
                    }
                    return@anyExecutor
                }

                world.difficulty = difficulty

                executor.sendText {
                    appendPrefix()
                    success("Du hast den Schwierigkeitsgrad der Welt ")
                    variableValue(world.name)
                    success(" auf ")
                    translatable(difficulty.translationKey()).colorIfAbsent(Colors.VARIABLE_VALUE)
                    success(" gesetzt.")
                }
            }
        }
    }
}